### ISSUE TITLE
Cannot copy Batchnorm weights, shape mismatch. 

### ISSUE THREAD

**Issue by WilliamJuel** [NVIDIA/caffe#307]
I have successfully trained a Resnet-50 from scratch on a 5 class problem, here is my prototxt [resnet50.txt](https://github.com/NVIDIA/caffe/files/929995/resnet50.txt)
I am on the nvidia caffe-0.15 branch, and i trained this network using DIGITS 5.

I want to use my trained caffemodel to calculate specificity and sensitivity on my test set (medical images), i have made this python script [infer_many.txt](https://github.com/NVIDIA/caffe/files/930010/infer_many.txt) to test a folder of images. (This works for me on other models that does not use batchnorm layers) 

The problem here is i, for some reason, cannot copy my batchnorm layer weights. I get the following Error:

"Cannot copy param 0 weights from layer 'xbn_conv1'; shape mismatch.  Source param shape is 1 64 1 1 (64); target param shape is 64 (64). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer. "

Obviously i could rename the layers, but this makes no sense since i do not want to retrain but rather use my already trained model to predict on a folder of test images.

Does anyone have a suggestion how to fix this mismatching problem ?

Heres is my deploy file [deploy.txt](https://github.com/NVIDIA/caffe/files/930042/deploy.txt)

**Comment by d-lareg** [comment]
Hello William, have you read this post?

[Incompatibilities in BatchNorm](https://github.com/NVIDIA/caffe/issues/276)

**Comment by WilliamJuel** [comment]
Helle @d-lareg thank you for replying, i did read it yes, but tbh i didnt expect that i couldnt use the net.cpp in caffe to copy my weights, even with nvidia caffe-0.15 batchnorm. Altough I found away around it by chaning the TrainedLayersFrom in net.cpp to this:

template <typename Dtype>
void Net<Dtype>::CopyTrainedLayersFrom(const NetParameter& param) {
  int num_source_layers = param.layer_size();
  for (int i = 0; i < num_source_layers; ++i) {
      const LayerParameter& source_layer = param.layer(i);
      const string& source_layer_name = source_layer.name();
      int target_layer_id = 0;
      while (target_layer_id != layer_names_.size() &&
          layer_names_[target_layer_id] != source_layer_name) {
          ++target_layer_id;
      }
      if (target_layer_id == layer_names_.size()) {
          LOG(INFO) << "Ignoring source layer " << source_layer_name;
          continue;
      }
      DLOG(INFO) << "Copying source layer " << source_layer_name;
      vector<shared_ptr<Blob<Dtype> > >& target_blobs =
          layers_[target_layer_id]->blobs();
      CHECK_EQ(target_blobs.size(), source_layer.blobs_size())
          << "Incompatible number of blobs for layer " << source_layer_name;
      //Modification to allow digits weights
      for (int j = 0; j < target_blobs.size(); ++j) {
          const bool bShapesAreEqual = target_blobs[j]->ShapeEquals(source_layer.blobs(j));
          const bool bSizeIsEqual = target_blobs[j]->count() == source_layer.blobs(j).data_size();
          if (bShapesAreEqual){
              const bool bReshape = false;
              target_blobs[j]->FromProto(source_layer.blobs(j), bReshape);
          }
          if (!bSizeIsEqual) {
              Blob<Dtype> source_blob;
              const bool kReshape = true;
              source_blob.FromProto(source_layer.blobs(j), kReshape);
              LOG(FATAL) << "Cannot copy param " << j << " weights from layer '"
                  << source_layer_name << "'; shape mismatch.  Source param shape is "
                  << source_blob.shape_string() << "; target param shape is "
                  << target_blobs[j]->shape_string() << ". "
                  << "To learn this layer's parameters from scratch rather than "
                  << "copying from a saved net, rename the layer.";
          }
          if (!bShapesAreEqual && bSizeIsEqual) {
              LOG(INFO) << "Shape is not equal but size is for layer " << source_layer_name;
              Blob<Dtype> source_blob;
              const bool bReshape = true;
              source_blob.FromProto(source_layer.blobs(j), bReshape);
              source_blob.ReshapeLike(*target_blobs[j]);
              target_blobs[j]->CopyFrom(source_blob);
          }
      }
  }
}

There, apparently, was some problems with the size of my mean file but i worked around that and it worked fine, and let my predict images. My problem now is that my code does not do the same predictions as Nvidia DIGITS does to the individual images, it is not completely wrong, but DIGITS does a better job.

**Comment by drnikolaev** [comment]
Thanks for reporting. Is this reproducible in caffe-0.16 branch?

**Comment by WilliamJuel** [comment]
Hey, I have not tried that. I have had some troubles with batchnorm layers and found out how it worked for the caffe-015 branch so i didnt want to change. Is it the same in the 0.16 branch? If so i could try. Does the net.cpp change from 0.15 to 0.16?

**Comment by WilliamJuel** [comment]
It just kinda sucks that i cant (atleast i cant figure out) how to deploy my caffemodel i get from training in DIGITS, without changing the net.cpp to the above. And when i do this i cannot reproduce the results from DIGITS. It would be nice if i could get a tip how i can deploy my model and use it outside of digits.

**Comment by d-lareg** [comment]
Hello again,
I read your deploy.txt and saw, that your batch size is 1. In this case, you better set...

use_global_stats: true

.., on every Batch Norm Layer. That would not solve the problem with your weights but might be the reason for the different results. Hope it helps. For more information please read:

[Batch Norm Layer](http://caffe.berkeleyvision.org/tutorial/layers/batchnorm.html)

**Comment by WilliamJuel** [comment]
@d-lareg hey, i actually have use_global_stats: true, i just saw i didnt change it in the deploy.txt i posted.

**Comment by drnikolaev** [comment]
0.16 supports both BN flavors and it actually autodetects which one is used in your model.

**Comment by WilliamJuel** [comment]
@drnikolaev Oh, that sounds smart. Would this fix my problem with not being able to deploy my caffemodel? I will try give it a try tomorrow tho. Thanks alot.

**Comment by WilliamJuel** [comment]
@drnikolaev I was (luckily) not able to reproduce the error with 0.16. Which mean that it works perfectly and i can deploy my model, thanks alot for the tip, great change :+1:

But i have three things:

1. I do not get the exact same results as in digits, on the same folder of images. There is a difference but it is small (i actually get better results). Is this expected?
2. When i crop my images in digits using the "crop_size: 224" and at the same time subtract image mean i get a shape mismatch when i deploy my model. Basically it says that my mean image does not fit with the input image. So i can sadly not use the  "crop_size:" if i want to deploy it. I have tried resizing the mean image but i get terrible results from the classification.
3. After i changed to 0.16 my val loss and val accuracy starts at 0, but after epoch 1 it goes to where it normally would be after 1 epoch. Normally the val loss started at the same place as the train loss and the val accuracy started at random (so 20% for this 5 class problem). What is going on here?
The image just shows my 5 class problem on a small net i used to test batchnorm on for the deployment. It is just a test net but it still shows the problem, i explained above.

![screenshot from 2017-04-27 17-47-14](https://cloud.githubusercontent.com/assets/26434742/25491931/c8fa6772-2b71-11e7-9742-291077219766.png)

Thanks for your time. Much appreciated!

**Comment by drnikolaev** [comment]
1. This may happen. Let me clarify:
1.1 NVCaffe 0.16 is deterministic as long as you use deterministic algorithms for convolution (for example, `default_conv_algos_override: "1,1,1"`) and you set the same random seed (like this: `random_seed: 1371`).
1.2 In our experiments we observed up to 0.5% of accuracy difference when we just change the seed only.

2. Please try to recompute (not to resize) the mean.

3. Could you please attach NVCaffe log? This would help a lot, thank you.

**Comment by d-lareg** [comment]
I assume that you are using use_global_stats=false during the whole training including the test phase? If you now use global stats for your deployed network your results will slightly differ.

**Comment by WilliamJuel** [comment]
@drnikolaev ahh i understand, thanks for clarifying.
Do you have a suggestion to how i can recompute it? 
Here it is. This is the log for a network im training atm. It has the same "problem" as shown on above image.
[caffe_output.txt](https://github.com/NVIDIA/caffe/files/962448/caffe_output.txt)

@d-lareg When training i use global_stats = false, to actually train it.. When i deploy i use_global_stats = true to use what it have learned. :)

Thanks for the respones.

**Comment by drnikolaev** [comment]
@WilliamJuel Thanks for the log. Could you pull the latest code and retry?

**Comment by WilliamJuel** [comment]
@drnikolaev I pulled the latest code and ran it again. 
The same thing happens as shown on the image above where the val loss and val accurracy starts at 0 and after the first snapshot goes too there "normal" values. Here is the new log
[caffe_output.txt](https://github.com/NVIDIA/caffe/files/963802/caffe_output.txt)


Another thing i see, this might be a digits problem rather than caffe, is that there is a gap in the learning rate graph, and there the training loss just explodes, but goes back to normal fast. If you look closely at the LR graph you can see the gap and at that gap on the loss accuracy graph the train loss goes high.

![lrhole](https://cloud.githubusercontent.com/assets/26434742/25520470/4f78dfba-2bfc-11e7-97c3-5edaf4fa7365.png)

**Comment by drnikolaev** [comment]
@lukeyeager could you check please?

**Comment by WilliamJuel** [comment]
I also noticed that i , after i switched to caffe-0.16, cannot reproduce my training in digits when setting a random seed.
Using:
Digits 5.1 and cuDNN 6.0

**Comment by RSly** [comment]
hi @WilliamJuel 
I also tried bn before but got many trouble.

could you please summerize here what deffinitions worked for nvcaffe 0.15 and 0.16? or even better upload your last proto?

thanks

**Comment by drnikolaev** [comment]
@WilliamJuel I forgot to mention one more condition: if you use Pooling you need to switch back to non-cuDNN implementation like this: `engine: CAFFE`. 
Corrected statement:
NVCaffe 0.16 is deterministic as long as you use deterministic algorithms for convolution (for example, `default_conv_algos_override: "1,1,1"`) and you set the same random seed (like this: `random_seed: 1371`) and you use deterministic Pooling algorithm (i.e. its non-CUDNN implementation: `engine: CAFFE`).

**Comment by WilliamJuel** [comment]
@drnikolaev So i need to use engine: CAFFE in my BN layers ? 

Like this:


```
layer {
	bottom: "conv1"
	top: "bn_conv1"
	name: "bn_conv1"
	type: "BatchNorm"
	 batch_norm_param {
    moving_average_fraction: 0.98
    use_global_stats: false
    eps: 0.0001
    scale_filler {
      type: "constant"
      value: 1 
    }
    bias_filler {
      type: "constant"
      value: 0 
    }
    engine: CAFFE
  }
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 1
    decay_mult: 1
  }  
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }  
}
```

**Comment by WilliamJuel** [comment]
@RSly From my understanding, and i think @drnikolaev  can correct me if im wrong. With Caffe 0.16 you can use BVLC/caffe implementation of BN and you can use what i wrote in my last post. With scale and bias in the BN layer instead of having an adjacent scale layer like in regular BVLC/caffe.

**Comment by RSly** [comment]
thanks @Williamjuel
I will try it! but Ithink the story of bn is far from converging.
see this also https://github.com/caffe2/caffe2/issues/435

**Comment by WilliamJuel** [comment]
@RSly Yeah well. Im sure they will implement it, so it auto detects if BN from caffe is used in a caffe 2 net, at some point. If you read, that is what drnikolav explained to me caffe 0.16 could do. But i agree, i have spend alot of time trying to train networks using BN. With little sucess. Maybe @drnikolaev could confirm if the prototxt example of a BN layer in my post just above is the correct way or not.

**Comment by drnikolaev** [comment]
@WilliamJuel No, not BN. I meant Pooling ( if we still discuss determinism). If your question is about auto-detection, then again, no need to switch the engine, keep it as cuDNN.

**Comment by WilliamJuel** [comment]
@drnikolaev okay thanks for clarifying, i get it! Could you confirm that the way i define the BN layer in the above post is a correct way to do it, obviously without the `engine: CAFFE` part i added?

I also noticed that since i pulled the latest 0.16 i cannot use my python layer that i have used in 0.15. i get this error from digits: ERROR: error code -11

and in the box it says:
Network initialization done.
Solver scaffolding done.
Starting Optimization
Solving
Learning Rate Policy: fixed
Iteration 0, Testing net (#0)
Ignoring source layer train-data
NVML initialized on thread 140484002506496
NVML succeeded to set CPU affinity on device 0 
[caffe_output.txt](https://github.com/NVIDIA/caffe/files/967475/caffe_output.txt)
[python_layer.txt](https://github.com/NVIDIA/caffe/files/967483/python_layer.txt)

**Comment by borisgin** [comment]
The following definition for BatchNorm layers is recommended for caffe-0.16:
layer {
  name: "bn"
  type: "BatchNorm"
  bottom: "conv"
  top: "bn"
  batch_norm_param {
    moving_average_fraction: 0.9
    eps: 0.0001
    scale_bias: true
  }
}
By default nvcaffe will use cuDNN engine if you built the code with cuDNN.
 
There is no need in defining  multiple lines with 
   param {
      lr_mult: 0
      decay_mult: 0
   }
caffe will take care of this .

**Comment by WilliamJuel** [comment]
Thanks for all the answers. I will close this and open a new one describing my problem with my python layer after i upgraded to 0.16. I did not find an answer to why there is a "gap" in the learning rate graph and if that is actually affecting the loss as visualized above.
